Fixing a number of a11y problems with design panel components

barklund opened the pull request:
## Summary

This fixes a number of design panel accessibility issues, including:

* Adds always-visible focus border to text inputs (both numeric and alphanumeric)
* Adds keyboard-only-visible focus border to drop downs, buttons and toggles
* Changes dropdown button to be an actual button
* Re-focus dropdown button when drop down closes when using keyboard (to not revert focus to canvas)
* Internal changes to padding control, so focus stays on the toggle when toggled (doesn't replace component)

## Relevant Technical Choices

* Some technical changes to color input structure and padding control necessitated by the above

## To-do

- [x] Improve keyboard accessibility for range slider (grid view, image editor, page advancement)
- [x] Improve keyboard accessibility for color input component (only design panel element, not color picker)
- [x] Improve keyboard accessibility for drop down (font family, font weight, author)
- [x] Improve keyboard accessibility for radio group (post status)
- [x] Improve keyboard accessibility for small toggle buttons (alignment, justification, bold/italic/underline, etc)
- [x] Improve keyboard accessibility for large toggle button/switch (video loop, auto page advancement)
- [x] Improve keyboard accessibility for number input component (multiple uses)
- [x] Improve keyboard accessibility for text input component (alt text, video title)
- [x] Improve keyboard accessibility for "edit color presets" button (keeps same button focused when pressed)
- [x] Improve keyboard accessibility for "toggle locked vertical/horizontal padding" button (keeps same button focused when pressed)
- [x] Move to new structure for annotating CUJ tests
- [x] Update existing tests to new CUJ test annotation structure
- [x] Add integration tests with screenshots of alignment panel
- [x] Add integration tests with screenshots of background overlay panel
- [x] Add integration tests with screenshots of link panel

## TODO in future PR's
- [ ] Add integration tests with screenshots of position & size panel
- [ ] Add integration tests with screenshots of background position & size panel
- [ ] Add integration tests with screenshots of shape style panel
- [ ] Add integration tests with screenshots of page style panel
- [ ] Add integration tests with screenshots of layer style panel
- [ ] Add integration tests with screenshots of image accessibility panel
- [ ] Add integration tests with screenshots of video accessibility panel
- [ ] Add integration tests with screenshots of video options panel
- [ ] Add integration tests with screenshots of text style panel
- [ ] Add integration tests with screenshots of style preset panel
- [ ] Add integration tests with screenshots of document page advancement panel
- [ ] Add integration tests with screenshots of document publish panel
- [ ] Add integration tests with screenshots of document slug panel
- [ ] Add integration tests with screenshots of document status panel

## User-facing changes

Compare how navigating with keyboard vs mouse looks now:

| With keyboard | With mouse
|-|-|
| ![designpanel_keyboard](https://user-images.githubusercontent.com/637548/86175582-9bae5680-baf1-11ea-9b98-7d3bba39d533.gif) | ![designpanel_mouse](https://user-images.githubusercontent.com/637548/86175601-a537be80-baf1-11ea-9311-7513890ff14d.gif) |
| All elements have focus border | Only inputs have focus border |

## Testing Instructions

Tab through fields and observe that focus is always visible.

---

<!-- Please reference the issue(s) this PR addresses. -->

Fixes #737
Fixes #2084

Commit:
A number of a11y fixes for design panel components

Commit:
Merge branch 'master' into fix/737-form-element-focus

Commit:
Merge branch 'master' into fix/737-form-element-focus

Commit:
Fixed default props for dropdown

Commit:
Make a bunch of buttons in design panel react to "enter"

Commit:
Add focus styling to resize handle (not the best, but better than nothing)

Commit:
Added a bunch of a11y improvements to preset panel

Commit:
Removed tab index from tabpanel, since all panels have focusable content

swissspidy commented on assets/src/edit-story/components/form/dropDown/index.js:
```diff
@@ -160,20 +186,12 @@ function DropDown({
   labelledBy: PropTypes.string,
 };
 
-DropDown.defaultProps = {
```
What's the reason for not using defaultProps anymore?

barklund replied:
This component was using a mix of `defaultProps` and default values in component definition, which is a mess!

I decided to switch to only the latter, as it seems to be the recommended approach these days (and [`defaultProps` will get deprecated at some point for functional components](https://twitter.com/dan_abramov/status/1133878326358171650?lang=en) anyway).

Commit:
Added a11y for radio group

Commit:
Added a11y for switch

Commit:
Added much better keyboard interaction and styling for range input

Commit:
Fix page advancement test

miina commented on assets/src/edit-story/components/form/radioGroup/index.js:
```diff
@@ -71,8 +72,14 @@ const Helper = styled.div`
   const radioGroupId = useMemo(() => uuidv4(), []);
+
+  // We need manual arrow key navigation here, as we have a global listener for those keys
+  // preventing default functionality.
```
Maybe we should instead just remove `radio` from the `NON_EDITABLE_INPUT_TYPES`? This would allow the default behavior which would make sense in case of radios. WDYT?

barklund replied:
Hm, not a bad idea, will look into it.

barklund replied:
So this actually didn't work - and the same for the buttons. Not completely sure why, maybe something about the events not bubbling up correctly.

miina replied:
Hmm, odd, when I test locally (on the main branch), it seems to work when just removing `radio` from `NON_EDITABLE_INPUT_TYPES`. E.g. I remove this, click on "Status and Visibility" Panel, tab into the radio button, use the arrow keys and the values change.

Is it not working at all when you try or it working in an unexpected way?

miina commented on assets/src/edit-story/components/rangeInput/index.js:
```diff
@@ -58,4 +64,52 @@ export default styled.input.attrs(({ min, max, step, value, onChange }) => ({
+
+function RangeInput({ minorStep, majorStep, handleChange, value, ...rest }) {
+  const ref = useRef();
+  const update = useCallback(
```
Is this respecting the min-max values?

barklund replied:
Good question!

miina commented on assets/src/edit-story/components/panels/textStyle/padding.js:
```diff
@@ -0,0 +1,162 @@
+    []
+  );
+
+  const firstInputProperties = lockPadding
```
Curious, what was the reason for combining the padding files? DRY? Or accessibility reasons, too?
It's reducing the readability a bit but perhaps the benefits have more weight.

barklund replied:
Accessibility - I wanted keyboard focus to stay on the lock icon button when clicking it (so when tabbing through and pressing enter on the button, you could keep tabbing from that same button). Thus it had to be a single component and had to return the same object, but with slightly different properties for the button, so React would treat it as the same DOM element and the browser would not move focus around.

I actually first tried manually refocusing the "new" button, but it was horrible.

miina commented on assets/src/edit-story/components/panels/stylePreset/header.js:
```diff
@@ -80,36 +107,35 @@ function PresetsHeader({
           >
-            <EditPencil />
-          </EditModeButton>
+            <AddPage />
```
`AddPage`? Let's rename to something more accurate, for example `AddPreset`?

Maybe generally the icon should be renamed to `Add` and then used as `Add*` as needed in different contexts? Thoughts?

barklund replied:
Didn't even consider that, but it makes sense. Will look into it.

miina commented on assets/src/edit-story/components/panels/stylePreset/header.js:
```diff
@@ -80,36 +107,35 @@ function PresetsHeader({
 
   const getActions = () => {
-    return !isEditMode ? (
+    return (
```
Same question here, was DRY the main reason for combining the two buttons into one, or also accessibility-related?

barklund replied:
Focus management again.

miina commented on assets/src/edit-story/components/rangeInput/index.js:
```diff
@@ -58,4 +64,52 @@ export default styled.input.attrs(({ min, max, step, value, onChange }) => ({
+  }
 `;
+
+function RangeInput({ minorStep, majorStep, handleChange, value, ...rest }) {
```
The `majorStep` is basically the default step, is that correct? The naming feels a bit odd as if there would be no option to do a "normal" step -- it's always either minor or major. WDYT if it would be just `step` and `minorStep`?
Or perhaps alternatively just leave a comment for clarity that `majorStep` is the default step?

barklund replied:
`majorStep` is the default when using keyboard, however when using the actual range, the "default" step (the smallest interval you can move with the mouse) would be the `minorStep`. I thought it was fairly self-explanatory but clearly not - will add some commentary.

miina replied:
Oh, right, I was only focusing on the keyboard part and didn't even think about the mouse while reviewing.

miina commented on assets/src/edit-story/components/panels/panel/shared/title.js:
```diff
@@ -106,6 +107,32 @@ const Collapse = styled.button`
 
+function Toggle({ children, toggle, ...rest }) {
+  // We unfortunately have to manually assign this listener, as it would be default behaviour
+  // if it wasn't for our listener further up the stack interpreting enter as "enter edit mode"
```
I wonder if instead, the "enter edit mode" could be adjusted so that it would be ignored when relevant input is in focus. E.g. in a similar way as `isEditableTarget` or `nativeCopyPasteExpected` are working. Thoughts?

barklund replied:
That would be an option too - but not easy to do. I don't think I'll look into that option in this PR, but it's definitely an option for improvement.

miina commented on assets/src/edit-story/components/form/switch.js:
```diff
@@ -120,9 +121,16 @@ function Switch({ value, disabled, onChange, onLabel, offLabel }) {
+  const ref = useRef();
+  useKeyDownEffect(
+    ref,
+    ['space', 'enter', 'left', 'right'],
```
Is it expected that if the value is already on the "right" then clicking `right` will still toggle it? Thoughts?

barklund replied:
Good question - I debated it with myself and thought that it didn't matter that this happened. It was slightly unexpected, but not a huge deal IMO.

Commit:
Merge branch 'master' into fix/737-form-element-focus

Commit:
Made range input arrow navigation respect min-max and added documentation

Commit:
Fix bug in color presets if adding text color and having gradient already

Commit:
Merge branch 'master' into fix/737-form-element-focus

Commit:
Renamed `AddPage` icon to `Add`

Commit:
Improved documentation for range input

Commit:
Merge branch 'master' into fix/737-form-element-focus

Commit:
Fixed keyboard focus design for color button for non-solid

Commit:
Use debounce when sending color picker change upstream

Commit:
Change color of focus border depending on current color luminance

Commit:
Fixed proptype for `Icon`

Commit:
Update tests to work with debounced color picker

Commit:
Fixed warning in karma tests about react-modal accessibility

Commit:
Fixed unit tests post color picker debounce

With one *todo* though, as we cannot wait for a negative result.

Commit:
Restructure how CUJ karma testing is annotated

Commit:
Fixed minor clarification in how CUJ is tallied

Commit:
Merge branch 'main' into fix/737-form-element-focus

Commit:
Updated tests to new structure

Commit:
Merge branch 'main' into fix/737-form-element-focus

Commit:
Move general custom matchers to global karma init

Commit:
Added (partial) test for alignment panel

It is only testing focus border for now, but has skipped tests for all other parts.

Commit:
Fixed karma CUJ aggregator

Commit:
Updated background overlay karma tests

Added relevant containers and added some screenshot tests for focus rects.

Commit:
Added tests (and disabled tests) for link panel

Commit:
Merge branch 'main' into fix/737-form-element-focus

Commit:
Fixed lints

Commit:
Updated dashboard karma tests to new CUJ annotations

merapi reviewed (commented):
Should this radio group have a focus border when navigating with the mouse?

![image](https://user-images.githubusercontent.com/21036224/89268286-9c7a6100-d638-11ea-88a1-4fe85008fa08.png)

merapi reviewed (commented):
If the first color is not applicable to text and you have text element selected - you will not be able to focus and change colors via keyboard.

Should we file a separate issue for it (since it was not introduced here) or it can be quickly fixed here?

![image](https://user-images.githubusercontent.com/21036224/89270125-2f1bff80-d63b-11ea-9a90-bfed3570259a.png)

merapi reviewed (approved):
Overall: great improvement, now the design panel is so much nicer when navigating with the keyboard ⌨️ 👍

Commit:
Merge branch 'main' into fix/737-form-element-focus

# Conflicts:
#	assets/src/edit-story/components/form/color/colorBox.js
#	assets/src/edit-story/components/panels/stylePreset/header.js

Commit:
Remove warning

Commit:
Merge branch 'main' into fix/737-form-element-focus

Commit:
Made radio group focus rect keyboard-only
